File: 034_find_first_and_last_position_of_element_in_sorted_array/Solution2/Solution.py
class Solution:
    def searchRange(self, nums: list, target: int) -> list:
        if not nums or len(nums) == 0: return [-1, -1]
        return [
            self.binary_search(nums, target, True),
            self.binary_search(nums, target, False)
        ]

    def binary_search(self, nums: list, target: int, is_search_first: bool):
        arr_len = len(nums)
        left, right = 0, arr_len - 1
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] > target: right = mid - 1
            elif nums[mid] < target: left = mid + 1
            else:
                if is_search_first:
                    if mid > 0 and nums[mid] == nums[mid - 1]: right = mid - 1
                    else: return mid
                else:
                    if mid < arr_len - 1 and nums[mid] == nums[mid + 1]: left = mid + 1
                    else: return mid

        return -1


if __name__ == '__main__':
    nums = [5, 7, 7, 8, 8, 10]
    target = 8
    solution = Solution()
    print(solution.searchRange(nums, target))

File: 034_find_first_and_last_position_of_element_in_sorted_array/Solution2/Solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        return new int[] {
            binarySearch(nums, target, true),
            binarySearch(nums, target, false)
        };
    }

    public int binarySearch(int[] nums, int target, boolean isSearchFirst) {
        int arrLen = nums.length;
        if (arrLen == 0) return -1;

        int left = 0, right = arrLen - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (nums[mid] < target) left = mid + 1;
            else if (nums[mid] > target) right = mid - 1;
            else {
                if (isSearchFirst) {
                    if (mid > 0 && nums[mid] == nums[mid - 1]) right = mid - 1;
                    else return mid;
                } else {
                    if (mid < arrLen - 1 && nums[mid] == nums[mid + 1]) left = mid + 1;
                    else return mid;
                }
            }
        }

        return -1;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{5, 7, 7, 8, 8, 10};
        int target = 8;
        Solution solution = new Solution();
        int[] result = solution.searchRange(nums, target);
        System.out.println(result[0] + " " + result[1]);
    }
}
